app/src/main/java/net/squanchy/schedule/filterschedule/FilterSchedule.kt

```diff
@@ -0,0 +1,143 @@
+        trackFiltersList.adapter = trackAdapter
+    }
+
+    override fun onStart() {
+        super.onStart()
+
+        subscription = Observable.combineLatest(tracksRepository.tracks(), tracksFilter.selectedTracks, combineIntoCheckableTracks())
+            .subscribeOn(Schedulers.computation())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe { checkableTracks ->
+                trackAdapter.updateTracks(checkableTracks) { track, selected ->
+                    val selectedTracks = checkableTracks.allSelected()
+                    val newSelectedTracks = selectedTracks.addOrRemove(track, selected)
+                    tracksFilter.updateSelectedTracks(newSelectedTracks)
+                    trackAdapter.notifyItemChanged(checkableTracks.map { checkableTrack -> checkableTrack.first }.indexOf(track))
+                }
+            }
+    }
+
+    private fun Iterable<CheckableTrack>.allSelected(): Set<Track> =
+        filter { it.selected() }.map { it.track() }.toSet()
+
+    private fun Set<Track>.addOrRemove(track: Track, selected: Boolean): Set<Track> =
+        if (selected) this + track else this - track
+
+    private fun combineIntoCheckableTracks(): BiFunction<List<Track>, Set<Track>, List<CheckableTrack>> {
+        return BiFunction { tracks, selectedTracks ->
+            tracks.map { track -> CheckableTrack(track, selectedTracks.contains(track)) }
+        }
+    }
+
+    override fun onStop() {
+        super.onStop()
+        subscription?.dispose()
+    }
+}
+
+private class TrackFiltersAdapter(context: Context) : RecyclerView.Adapter<TrackViewHolder>() {
+
+    init {
+        setHasStableIds(true)
+    }
+
+    private val layoutInflater = LayoutInflater.from(context)
+
+    private var checkableTracks: List<CheckableTrack> = emptyList()
+    private lateinit var trackStateChangeListener: OnTrackSelectedChangeListener
+
+    override fun getItemCount(): Int = checkableTracks.size
+
+    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): TrackViewHolder {
+        val view = layoutInflater.inflate(R.layout.track_filters_item, parent, false) as CheckBox
+        return TrackViewHolder(view)
+    }
+
+    override fun onBindViewHolder(holder: TrackViewHolder, position: Int) {
+        holder.bind(checkableTracks[position], trackStateChangeListener)
+    }
+
+    override fun getItemId(position: Int): Long {
+        return checkableTracks[position].track().id.hashCode().toLong() // TODO this should use a proper checksum
+    }
+
+    fun updateTracks(newCheckableTracks: List<CheckableTrack>, listener: OnTrackSelectedChangeListener) {
+        checkableTracks = newCheckableTracks // TODO use DiffUtil instead
+        trackStateChangeListener = listener
+        notifyDataSetChanged()
+    }
+}
+
+private class TrackViewHolder(val item: CheckBox) : RecyclerView.ViewHolder(item) {
+
+    fun bind(checkableTrack: CheckableTrack, listener: OnTrackSelectedChangeListener?) {
```

Comment:
this can now be with `OnTrackSelectedChangeListener` (without the `?`)

Reply:
999cf6546e345386a62ae1417ffaaa54aa61497d